ras_party/controllers/respondent_controller.py

```diff
@@ -125,17 +125,63 @@ def update_respondent_mark_for_deletion(email: str, session):
 
     :param session A db session
     """
     respondents = session.query(Respondent).filter(Respondent.mark_for_deletion == True)  # noqa
+    logger.info("Preparing to delete respondent records", number_to_delete=respondents.count())
+    failed_deletion_count = 0
     for respondent in respondents:
-        session.query(Enrolment).filter(Enrolment.respondent_id == respondent.id).delete()
-        session.query(BusinessRespondent).filter(BusinessRespondent.respondent_id == respondent.id).delete()
-        session.query(PendingEnrolment).filter(PendingEnrolment.respondent_id == respondent.id).delete()
-        session.query(Respondent).filter(Respondent.id == respondent.id).delete()
-        send_account_deletion_confirmation_email(respondent.email_address, respondent.first_name)
+        bound_logger = logger.bind(
+            email=obfuscate_email(respondent.email_address),
+            respondent_id=respondent.id,
+            party_uuid=respondent.party_uuid,
+        )
+        bound_logger.info("Attempting to delete respondent records")
+        try:
+            delete_respondent_records(session, respondent)
+            send_account_deletion_confirmation_email(respondent.email_address, respondent.first_name)
+            bound_logger.info("Respondent records deleted successfully")
+        except IntegrityError as e:
+            logger.error(
+                "A data constraint violation occurred trying to delete the respondent records",
+                respondent_id=respondent.id,
+                party_uuid=respondent.party_uuid,
+                error=str(e),
+            )
+            session.rollback()
+            failed_deletion_count += 1
+        except SQLAlchemyError as e:
+            logger.error(
+                "An error occurred trying to delete the respondent records",
+                respondent_id=respondent.id,
+                party_uuid=respondent.party_uuid,
+                error=str(e),
+            )
+            session.rollback()
+            failed_deletion_count += 1
+    logger.info("Respondent record deletions complete", failed_deletion_count=failed_deletion_count)
+
+def delete_respondent_records(session, respondent):
+    """
+    Deletes all records associated with a respondent.
+
+    :param session: The transactional db session that can be rolled back
+    :param respondent: The respondent whose records are to be deleted
+    :type respondent: Respondent
+    :raises IntegrityError: If there is a data constraint violation
+    :raises SQLAlchemyError: If there is a general SQLAlchemy error
+    """
+    session.query(Enrolment).filter(Enrolment.respondent_id == respondent.id).delete()
+    session.query(BusinessRespondent).filter(BusinessRespondent.respondent_id == respondent.id).delete()
+    session.query(PendingEnrolment).filter(PendingEnrolment.respondent_id == respondent.id).delete()
+    session.query(Respondent).filter(Respondent.id == respondent.id).delete()
+    session.commit()
```

Comment:
If the above deletions are successful this commit gets actioned, thus ensuring that respondent is really deleted regardless of any upcoming failures for other respondents.